reana_server/gitlab_client.py

```diff
@@ -0,0 +1,266 @@
+# This file is part of REANA.
+# Copyright (C) 2024 CERN.
+#
+# REANA is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+"""REANA-Server GitLab client."""
+
+from typing import Dict, Optional, Union
+from urllib.parse import quote_plus
+import requests
+import yaml
+
+from reana_commons.k8s.secrets import REANAUserSecretsStore
+
+from reana_server.config import REANA_GITLAB_HOST
+
+
+class GitLabException(Exception):
+    """Base class for GitLab exceptions."""
+
+    def __init__(self, message):
+        """Initialise the GitLabException exception."""
+        self.message = message
+
+    def __str__(self):
+        """Return the exception message."""
+        return self.message
+
+
+class GitLabRequestError(GitLabException):
+    """Raised when a GitLab API request fails."""
+
+    def __init__(self, response, message=None):
+        """Initialise the GitLabRequestError exception."""
+        message = message or f"GitLab API request failed: {response.status_code}"
+        super().__init__(message)
+        self.response = response
+
+
+class InvalidGitLabTokenError(GitLabException):
+    """Raised when GitLab token is invalid or missing."""
+
+    def __init__(self, message=None):
+        """Initialise the InvalidGitLabTokenError exception."""
+        message = message or (
+            "GitLab token invalid or missing, "
+            "please go to your profile page on REANA "
+            "and reconnect to GitLab."
+        )
+        super().__init__(message)
+
+
+class GitLabClient:
```

Comment:
Hmm, is there some existing GitLab client library which we  could perhaps adopt, such as [python-gitlab](https://python-gitlab.readthedocs.io/en/stable/)? We don't need much, so we can go with our own. Just wondering for future maintainability purposes, possible versioning when GitLab upgrades, etc.

Reply:
In any case, we would probably still need to wrap the library to use our custom errors, or we would need to change the rest of the code to check for the library's errors.